### ISSUE TITLE
TfLite on object detection with QUANTIZED_UINT8.

### ISSUE THREAD

**Issue by jiarenyf** [tensorflow/tensorflow#26749]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 18.04.
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: XiaoMi 8.
- **TensorFlow installed from (source or binary)**: pip install tensorflow-gpu==1.9.0
- **TensorFlow version (use command below)**: 1.9.0
- **Python version**: 2.7.12
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: 9.0/7.1
- **GPU model and memory**: enough.
- **Exact command to reproduce**:

### Describe the problem
I train a model with only multi-layer conv+bn+relu/sigmoid using `tf.contrib.quantize.create_training_graph()` and test it with `tf.contrib.quantize.create_eval_graph()`, every seems ok.

Then I freeze the *.ckpt file to *.pb file with `tensorflow.python.tools.freeze_graph.freeze_graph`.
Finally, I convert the *.pb file to *.tflite file with the following code, and again every is ok.

```
    lite8 = tf.contrib.lite.toco_convert(
        input_tensors=[inT],
        output_tensors=[outT],
        input_data=graph.as_graph_def(),
        # default_ranges_stats=(-6.0, 6.0),
        quantized_input_stats=[(127.5, 128.0)],
        # inference_type=tf.contrib.lite.constants.FLOAT,
        inference_type=tf.contrib.lite.constants.QUANTIZED_UINT8,
    )

    with open(lite8Path, 'wb') as txt:
        txt.write(lite8)
```

**Here is the question:**
The output of my model is the concat of 2 tensors, the first is the output of `cnn+bn`, the second one is the output of `cnn+bn+sigmoid`, **both should be type of float32**. However, when I run the quantized model on android phone with the following code, I am force to get output tensor of type of `uint8` rather than 'float32'.

```
        byte outB[][][] = new byte[1][NUM * NUM][2 + 4]; # The output is a tensor of rank 3.
        org.tensorflow.lite.Interpreter.run(input_value, outB); # The outB must be uint8 (byte in android) here.

        float outF[][][] = new Float[1][NUM * NUM][2 + 4];
        org.tensorflow.lite.Interpreter.run(input_value, outF); # The application crash in this line, saying the output dtype is not match.
```

**I wonder how I can get `float32` output rather than `uint8` when using quantized .tflite file ?**
**Thank you very much !!!**

**In short, I have convert a pb file to tflite with uint8 quantization. I would like to know how to convert the output from uint8 back to float so that the previous code (that work on float type output) can still work.**

### Source code / logs
Clear enough, need not to code.

**Comment by jiarenyf** [comment]
@jdduke 
@andrehentz  
@shashishekhar 

Help me, please !!!

**Comment by shashishekhar** [comment]
@jiarenyf : 
Right now you can only convert the entire graph to uint8. One way you can avoid quantizing a particular layer would be changing the output_tensors specified in toco_convert, but then you may need to add some custom post processing for the model. Since the graph will only be converted till the output tensor.
Adding @suharshs who may be able to give a definitive answer.

**Comment by jiarenyf** [comment]
@shashishekhar Thank you.

In short, I have convert a PB file to tflite with uint8 quantization. I would like to know how to convert the output from uint8 back to float so that the previous code (that work on float type output) can still work.

@suharshs Could you please help me ?

**Comment by jiarenyf** [comment]
@shafi-dayatar Finally I have struggled to find the solution:

**First, read the quantized_min/max value from the ckpt file with the output tensor names:**
```
import tensorflow as tf

folder = './model8'
iCheckpoint = tf.train.latest_checkpoint(folder)
reader = tf.train.NewCheckpointReader(iCheckpoint)

vs_map = reader.get_variable_to_shape_map()
print(reader.get_tensor('conv2d_11/act_quant/min')) # As `MIN_1` ...
print(reader.get_tensor('conv2d_11/act_quant/max')) # As `MAX_1` ...
print(reader.get_tensor('conv2d_12/act_quant/min')) # As `MIN_2` ...
print(reader.get_tensor('conv2d_12/act_quant/max')) # As `MAX_2` ...
```
**And remember not to package the `sigmoid` op in the pb/tflite files, since it bring slide effect in the tflite interpreter.**

**Second, calculate the float32 using unit8 with min/max value:**
```
    private static final float MAX_1 = 28.67623f;
    private static final float MIN_1 = -28.629274f;
    private static final float SCALE_1 = (float) ((MAX_1 - MIN_1) / 255.0);

    private static final float MAX_2 = 4.380291f;
    private static final float MIN_2 = -5.4539404f;
    private static final float SCALE_2 = (float) ((MAX_2 - MIN_2) / 255.0);

    Bitmap nBitmap = Bitmap.createScaledBitmap(
                bitmap, IMG_WIDTH, IMG_HEIGHT, true);
        int intValues[] = new int[IMG_WIDTH * IMG_HEIGHT];
        nBitmap.getPixels(intValues, 0, IMG_WIDTH, 0, 0, IMG_WIDTH, IMG_HEIGHT);

        byte[][][][] bValues = new byte[1][IMG_HEIGHT][IMG_WIDTH][1];
        for (int i = 0; i < IMG_HEIGHT; ++i) {
            for (int k = 0; k < IMG_WIDTH; ++k) {
                int val = intValues[i * IMG_WIDTH + k];
                int b = val & 0xFF;
                int g = (val >> 8) & 0xFF;
                int r = (val >> 16) & 0xFF;

                bValues[0][i][k][0] = (byte) (r * 0.30 + g * 0.59 + b * 0.11);
            }
        }
        byte outB1[][][] = new byte[1][NUM * NUM][2];
        byte outB2[][][] = new byte[1][NUM * NUM][4];

        Object[] inputArray = {bValues};
        @SuppressLint("UseSparseArrays")
        Map<Integer, Object> outputMap = new HashMap<>();
        outputMap.put(0, outB1);
        outputMap.put(1, outB2);
        lite.runForMultipleInputsOutputs(inputArray, outputMap);

        float out1[][][] = new float[1][NUM * NUM][2];
        float out2[][][] = new float[1][NUM * NUM][4];
        for (int i = 0; i < NUM * NUM; ++i) {
            for (int k = 0; k < 2; ++k) {
                out1[0][i][k] = (outB1[0][i][k] & 0xFF) * SCALE_1 + MIN_1;
            }
            for (int k = 0; k < 4; ++k) {
                out2[0][i][k] = (outB2[0][i][k] & 0xFF) * SCALE_2 + MIN_2;
                out2[0][i][k] = (float) (1 / (1 + Math.exp(-out2[0][i][k])));
            }
        }
```
** We store the `uint8` values in `outB1` and `outB2` arrays, and calculate the corresponding `float32` values using the `MIN/MAX` values we got in Step 1.**

That all~
